include/boost/math/interpolators/cubic_hermite.hpp

```diff
@@ -0,0 +1,46 @@
+// Copyright Nick Thompson, 2020
+// Use, modification and distribution are subject to the
+// Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt
+// or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_MATH_INTERPOLATORS_PCHIP_HPP
+#define BOOST_MATH_INTERPOLATORS_PCHIP_HPP
+#include <memory>
+#include <boost/math/interpolators/detail/cubic_hermite_detail.hpp>
+
+namespace boost::math::interpolators {
+
+template<class RandomAccessContainer>
+class cubic_hermite {
+public:
+    using Real = typename RandomAccessContainer::value_type;
+
+    cubic_hermite(RandomAccessContainer && x, RandomAccessContainer && y, RandomAccessContainer && dydx) 
+    : impl_(std::make_shared<detail::cubic_hermite_detail<RandomAccessContainer>>(std::move(x), std::move(y), std::move(dydx)))
+    {}
+
+    Real operator()(Real x) const {
+        return impl_->operator()(x);
+    }
+
+    Real prime(Real x) const {
+        return impl_->prime(x);
+    }
+
+    friend std::ostream& operator<<(std::ostream & os, const cubic_hermite & m)
```

Comment:
I notice that `std::ostream` is used here but no i/o header is included above?
As a side note, do we not typically put i/o functions in separate headers so that users can avoid including them altogether if they wish?

Reply:
It's in the `boost/math/interpolators/detail/cubic_hermite_detail.hpp`.

Reply:
I didn't know of any such rule; how helpful would it be? We'd lose helpful debug info.

Reply:
> It's in the `boost/math/interpolators/detail/cubic_hermite_detail.hpp`.

Some people consider it a pedantic physical design rule, but the point is about dependency: this header now depends on the detail header to include a standard header for it. And for someone editing the detail header, they couldn't know how many other headers are depending on it to do that. So the rule is: don't depend on other headers to include things for you.  :)

Reply:
> I didn't know of any such rule; how helpful would it be? We'd lose helpful debug info.

It's not a rule, not a written one anyway, just something I have seen done elsewhere. Boost.Tuple, I think, and that's kinda old now, so maybe this idea doesn't apply anymore.
The point is not to lose anything, but just to make it optional, so that people who don't need it don't have to compile it.
I'm not sure what `std::ostream` brings in, maybe it's just an interface, but for example on many embedded targets, including `<iostream>` is strictly forbidden because of the extra ROM space it takes up. Some people just want to compute, and they don't care about the debugging.